server/repository/src/db_diesel/number_row.rs

```diff
@@ -55,6 +88,68 @@ impl<'a> NumberRowRepository<'a> {
+
+        // Note: Format string is used here because diesel does seems to support binding NumberRowType as a String or as it's own type.
+        // It's safe to use format here, as r#type is an enum with predefined values (No user input)
+        let update_query_str = format!(
+            r#"UPDATE number SET value = value+1 WHERE store_id = $1 and type = '{}' RETURNING value;"#,
+            r#type
+        );
+        let update_query = sql_query(update_query_str.clone()).bind::<Text, _>(store_id);
+
+        // Debug diesel query
+        // println!(
+        //     "{}",
+        //     diesel::debug_query::<crate::DBType, _>(&update_query).to_string()
+        // );
+        let update_result = update_query.get_result::<NextNumber>(&self.connection.connection);
+
+        match update_result {
+            Ok(result) => Ok(result),
+            Err(NotFound) => {
+                // 2. There was no record to update, so we need to insert a new one.
+
+                // We need to add an ON CONFLICT Clause for postgres just in case 2 threads insert at the same time (SQLite <on disk> does not need this as it only allows a single write transaction at a time).
+                // Without this postgres will throw a unique constraint violation error and rollback the transaction, which is hard to recover from, instead we just ignore the error and check if it returned a value
+                // It's safe to use format here, as these inputs are not user controlled
+                let insert_query_str = format!(
+                    r#"INSERT INTO number (id, value, store_id, type) VALUES ('{}', 1, $1, '{}') {} RETURNING value;"#,
+                    uuid(),
+                    r#type,
+                    ON_CONFLICT_DO_NOTHING
+                );
+
+                let insert_query = sql_query(insert_query_str).bind::<Text, _>(store_id);
+                let insert_result =
+                    insert_query.get_result::<NextNumber>(&self.connection.connection);
+
+                match insert_result {
+                    Ok(result) => Ok(result),
+                    Err(NotFound) => {
+                        // 3. If we got here another thread inserted the record before we we able to (we know this because nothing was returned for the insert)
+                        // We should now be able to do the same 'update returning' query as before to get our new number.
+
+                        let update_query =
+                            sql_query(update_query_str.clone()).bind::<Text, _>(store_id);
+
+                        match update_query.get_result::<NextNumber>(&self.connection.connection) {
+                            Ok(result) => Ok(result),
+                            Err(e) => Err(RepositoryError::from(e)),
+                        }
+                    }
+                    Err(e) => Err(RepositoryError::from(e)),
+                }
+            }
+            Err(e) => Err(RepositoryError::from(e)),
```

Comment:
Do you want to add a test case for all these 3 cases?

Reply:
I believe the test case provided tests all 3 cases. Let me know if you think something is missed.

Reply:
I added some debug output and the "insert new row conflict" (second error) is not reached. However, difficult to trigger in a test since you would need to pause in the repository method...

Reply:
Interesting it's definitely being hit sometimes in my test, are you testing with postgres or sqlite? SQL doesn't have this issue due to the single write transaction guarantee.

Reply:
tested with postgres...

Reply:
```
PART 1
PART 2
PART 1
PART 2
PART 3
next_number (INSERT) results : a=1 b=2
PART 1
PART 1
next_number (UPDATE) results : a=3 b=4
test number::test::test_concurrent_next_number ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 151 filtered out; finished in 0.81s
```

Here's what I get on my M1 Mac. Maybe it's because my postgres is a bit slow in docker?